sweep/sweeper.go

```diff
@@ -29,13 +64,387 @@ type UtxoSweeperConfig struct {
+		pendingInputs: make(map[wire.OutPoint]*pendingInput),
+	}
+}
+
+// Start starts the process of constructing and publish sweep txes.
+func (s *UtxoSweeper) Start() error {
+
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	log.Debugf("Starting sweeper")
+
+	s.wg.Add(1)
+	go s.collector()
+
+	return nil
+}
+
+// Stop stops sweeper from listening to block epochs and constructing sweep
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Infof("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Infof("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	log.Debug("Pushing new input to collector")
+
+	sweeperInput := &sweeperInput{
+		input:      input,
+		resultChan: make(chan Result),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+func (s *UtxoSweeper) collector() {
+	log.Debug("Collector started")
+	defer s.wg.Done()
+
+	s.cfg.Ticker.Resume()
+	defer s.cfg.Ticker.Stop()
+
+	inputToSweep := make(map[wire.OutPoint]Input)
+	clearInputsToSweep := func() {
+		inputToSweep = make(map[wire.OutPoint]Input)
+	}
+	sweepCountdown := 0
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners =
+					append(pendInput.listeners,
+						input.resultChan)
+
+				break
+			}
+
+			// Lookup if there is already a confirmed or unconfirmed
+			// tx of ours that spends this input.
+			spendingTx, err :=
+				s.cfg.HasSpendingTx(outpoint)
+			if err != nil {
+				s.signalResultChan(input.resultChan, Result{Err: err})
+				break
+			}
+
+			// Start watching for spend of this input, either by us
+			// or the remote party. The possibility of a remote
+			// spend is there reason why just registering for conf.
+			// of our sweep tx isn't enough.
+			cancel, err := s.waitForSpend(input.input)
+			if err != nil {
+				s.signalResultChan(input.resultChan,
+					Result{Err: fmt.Errorf(
+						"cannot watch for spend: %v",
+						err)})
+
+				break
+			}
+
+			// Create a new pendingInput struct and initialize the
+			// listeners slice with the passed in result channel. If
+			// this input is offered for sweep again, the result
+			// channel will be appended to this slice.
+			pendInput = &pendingInput{
+				listeners:     []chan Result{input.resultChan},
+				ntfnRegCancel: cancel,
+			}
+
+			if spendingTx != nil {
```

Comment:
If it's already been swept, then shouldn't be report back to the requester immediately?

Reply:
`spendingTx` can still be unconfirmed or confirmed with not enough confs. That's why we wait for a spend ntfn with enough depth. If there is a way to get the actual number of confs for a tx synchronously, we could short circuit here. But I don't think it is worth it.